pkg/controller/networkpolicy/networkpolicy_controller_perf_test.go

```diff
@@ -0,0 +1,273 @@
+				Spec:       v1.PodSpec{NodeName: getRandomNodeName()},
+				Status:     v1.PodStatus{PodIP: getRandomIP()},
+			},
+			{
+				ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: "pod2", UID: types.UID(uuid.New().String()), Labels: map[string]string{"app-1": "scale-1"}},
+				Spec:       v1.PodSpec{NodeName: getRandomNodeName()},
+				Status:     v1.PodStatus{PodIP: getRandomIP()},
+			},
+			{
+				ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: "pod3", UID: types.UID(uuid.New().String()), Labels: map[string]string{"app-2": "scale-2"}},
+				Spec:       v1.PodSpec{NodeName: getRandomNodeName()},
+				Status:     v1.PodStatus{PodIP: getRandomIP()},
+			},
+			{
+				ObjectMeta: metav1.ObjectMeta{Namespace: namespace, Name: "pod4", UID: types.UID(uuid.New().String()), Labels: map[string]string{"app-2": "scale-2"}},
+				Spec:       v1.PodSpec{NodeName: getRandomNodeName()},
+				Status:     v1.PodStatus{PodIP: getRandomIP()},
+			},
+		}
+		return namespaces, networkPolicies, pods
+	}
+	namespaces, networkPolicies, pods := getXObjects(25000, getObjects)
+	testComputeNetworkPolicy(t, 10*time.Second, namespaces, networkPolicies, pods)
+}
+
+func testComputeNetworkPolicy(t *testing.T, maxExecutionTime time.Duration, namespaces []*v1.Namespace, networkPolicies []*networkingv1.NetworkPolicy, pods []*v1.Pod) {
+	objs := make([]runtime.Object, 0, len(namespaces)+len(networkPolicies)+len(pods))
+	for i := range namespaces {
+		objs = append(objs, namespaces[i])
+	}
+	for i := range networkPolicies {
+		objs = append(objs, networkPolicies[i])
+	}
+	for i := range pods {
+		objs = append(objs, pods[i])
+	}
+
+	_, c := newController(objs...)
+	c.heartbeatCh = make(chan heartbeat, 1000)
+
+	var wg sync.WaitGroup
+	stopCh := make(chan struct{})
+
+	// executionMetric is used to count the executions of each routine and to record the last execution time.
+	type executionMetric struct {
+		executions    int
+		lastExecution time.Time
+	}
+	executionMetrics := map[string]*executionMetric{}
+
+	// If we don't receive any heartbeat from NetworkPolicyController for 3 seconds, it means all computation
+	// finished 3 seconds ago.
+	idleTimeout := 3 * time.Second
+	timer := time.NewTimer(idleTimeout)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			timer.Reset(idleTimeout)
+			select {
+			case heartbeat := <-c.heartbeatCh:
+				m, ok := executionMetrics[heartbeat.name]
+				if !ok {
+					m = &executionMetric{}
+					executionMetrics[heartbeat.name] = m
+				}
+				m.executions++
+				m.lastExecution = heartbeat.timestamp
+			case <-timer.C:
+				// Send the stop signal if we don't receive any heartbeat for 3 seconds.
+				close(stopCh)
+				return
+			}
+		}
+	}()
+
+	// maxAlloc is used to record the maximum heap allocation.
+	var maxAlloc uint64
+	var memStats goruntime.MemStats
+	wg.Add(1)
+	go func() {
```

Comment:
Scratch that. It would only apply if this was the only test we ran... Another test ran in the same process may have a larger memory footprint.

Maybe still move this code to a separate function for readability?

Reply:
Done